File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.java
package me.zeroeightsix.kami.module.modules.chat;

import me.zero.alpine.listener.EventHandler;
import me.zero.alpine.listener.Listener;
import me.zeroeightsix.kami.KamiMod;
import me.zeroeightsix.kami.event.events.PacketEvent;
import me.zeroeightsix.kami.module.Module;
import me.zeroeightsix.kami.setting.Setting;
import me.zeroeightsix.kami.setting.Settings;
import net.minecraft.network.play.server.SPacketChat;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import static me.zeroeightsix.kami.KamiMod.MODULE_MANAGER;
import static me.zeroeightsix.kami.util.MessageDetectionHelper.isDirect;
import static me.zeroeightsix.kami.util.MessageDetectionHelper.isDirectOther;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendChatMessage;
import static me.zeroeightsix.kami.util.MessageSendHelper.sendRawChatMessage;

/**
 * @author hub
 * @author dominikaaaa
 * Created 19 November 2019 by hub
 * Updated 12 January 2020 by hub
 * Updated 19 February 2020 by aUniqueUser
 * Updated by dominikaaaa on 19/04/20
 */
@Module.Info(
        name = "AntiSpam",
        category = Module.Category.CHAT,
        description = "Removes spam and advertising from the chat",
        showOnArray = Module.ShowOnArray.OFF
)
public class AntiSpam extends Module {

    private Setting<Page> p = register(Settings.e("Page", Page.ONE));
    /* Page One */
    private Setting<Boolean> discordLinks = register(Settings.booleanBuilder("Discord Links").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> announcers = register(Settings.booleanBuilder("Announcers").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> spammers = register(Settings.booleanBuilder("Spammers").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> insulters = register(Settings.booleanBuilder("Insulters").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> greeters = register(Settings.booleanBuilder("Greeters").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> ips = register(Settings.booleanBuilder("Server Ips").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> specialCharEnding = register(Settings.booleanBuilder("Special Ending").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> specialCharBegin = register(Settings.booleanBuilder("Special Begin").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    private Setting<Boolean> iJustThanksTo = register(Settings.booleanBuilder("I just...thanks to").withValue(true).withVisibility(v -> p.getValue().equals(Page.ONE)).build());
    /* I can't get settings to work in non static context for filter */
//    private Setting<Integer> characters = register(Settings.integerBuilder("Characters").withValue(15).withVisibility(v -> wordsLongerThen.getValue() && p.getValue().equals(Page.ONE)).build());

    /* Page Two */
    private Setting<Boolean> ownsMeAndAll = register(Settings.booleanBuilder("Owns Me And All").withValue(true).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> greenText = register(Settings.booleanBuilder("Green Text").withValue(false).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> numberSuffix = register(Settings.booleanBuilder("Number Ending").withValue(true).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> numberPrefix = register(Settings.booleanBuilder("Number Begin").withValue(false).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> duplicates = register(Settings.booleanBuilder("Duplicates").withValue(true).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Integer> duplicatesTimeout = register(Settings.integerBuilder("Duplicates Timeout").withMinimum(1).withValue(30).withMaximum(600).withVisibility(v -> duplicates.getValue() && p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> webLinks = register(Settings.booleanBuilder("Web Links").withValue(false).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> filterOwn = register(Settings.booleanBuilder("Filter Own").withValue(false).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<Boolean> filterDMs = register(Settings.booleanBuilder("Filter DMs").withValue(false).withVisibility(v -> p.getValue().equals(Page.TWO)).build());
    private Setting<ShowBlocked> showBlocked = register(Settings.enumBuilder(ShowBlocked.class).withName("Show Blocked").withValue(ShowBlocked.LOG_FILE).withVisibility(v -> p.getValue().equals(Page.TWO)).build());

    private ConcurrentHashMap<String, Long> messageHistory;
    private enum Page { ONE, TWO }
    private enum ShowBlocked { NONE, LOG_FILE, CHAT }

    @EventHandler
    public Listener<PacketEvent.Receive> listener = new Listener<>(event -> {
        if (mc.player == null || isDisabled()) return;
        if (!(event.getPacket() instanceof SPacketChat)) return;

        SPacketChat sPacketChat = (SPacketChat) event.getPacket();

        // servers i test on did not send ChatType.CHAT for chat messages >:(
        /*if (!sPacketChat.getType().equals(ChatType.CHAT)) {
            return;
        }*/

        /* leijurv's sexy lambda to remove older entries in messageHistory */
        messageHistory.entrySet()
                .stream()
                .filter(entry -> entry.getValue() < System.currentTimeMillis() - 10 * 60 * 1000) // 10 is delay in minutes
                .collect(Collectors.toList())
                .forEach(entry -> messageHistory.remove(entry.getKey()));

        String message = sPacketChat.getChatComponent().getUnformattedText();

        if (!isSpam(message)) {
            if (MODULE_MANAGER.isModuleEnabled(ChatTimestamp.class)) {
                message = MODULE_MANAGER.getModuleT(ChatTimestamp.class).getFormattedTime(message);
            }
            sendRawChatMessage(message);
        }
        event.cancel();
    });

    @Override
    public void onEnable() {
        messageHistory = new ConcurrentHashMap<>();
    }

    @Override
    public void onDisable() {
        messageHistory = null;
    }

    private boolean isSpam(String message) {
        /* Quick bandaid fix for mc.player being null when the module is being registered, so don't register it with the map */
        final String OWN_MESSAGE = "^<" + mc.player.getName() + "> ";
        if ((!filterOwn.getValue() && isOwn(OWN_MESSAGE, message)) || isDirect(filterDMs.getValue(), message) || isDirectOther(filterDMs.getValue(), message)) {
            return false;
        } else {
            return detectSpam(removeUsername(message));
        }
    }


    private String removeUsername(String username) {
        return username.replaceAll("<[^>]*> ", "");
    }

    private boolean detectSpam(String message) {

        for (Map.Entry<Setting<Boolean>, String[]> entry : settingMap.entrySet()) {
            if (entry.getKey().getValue() && findPatterns(entry.getValue(), message)) {
                sendResult(entry.getKey().getName(), message);
                return true;
            }
        }

        if (duplicates.getValue()) {
            if (messageHistory == null) messageHistory = new ConcurrentHashMap<>();
            boolean isDuplicate = false;

            if (messageHistory.containsKey(message) && (System.currentTimeMillis() - messageHistory.get(message)) / 1000 < duplicatesTimeout.getValue())
                isDuplicate = true;

            messageHistory.put(message, System.currentTimeMillis());
            if (isDuplicate) {
                if (showBlocked.getValue().equals(ShowBlocked.CHAT))
                    sendChatMessage(getChatName() + "Duplicate: " + message);
                else if (showBlocked.getValue().equals(ShowBlocked.LOG_FILE))
                    KamiMod.log.info(getChatName() + "Duplicate: " + message);
            }
        }

        return false;
    }

    private boolean isOwn(String ownFilter, String message) {
        return Pattern.compile(ownFilter, Pattern.CASE_INSENSITIVE).matcher(message).find();
    }

    private boolean findPatterns(String[] patterns, String string) {
        string = string.replaceAll("<[^>]*> ", ""); // remove username first
        for (String pattern : patterns) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(string).find()) {
                return true;
            }
        }
        return false;
    }

    private Map<Setting<Boolean>, String[]> settingMap = new HashMap<Setting<Boolean>, String[]>() {{
        put(greenText, FilterPatterns.GREEN_TEXT);
        put(specialCharBegin, FilterPatterns.SPECIAL_BEGINNING);
        put(specialCharEnding, FilterPatterns.SPECIAL_ENDING);
        put(specialCharBegin, FilterPatterns.SPECIAL_BEGINNING);
        put(ownsMeAndAll, FilterPatterns.OWNS_ME_AND_ALL);
        put(iJustThanksTo, FilterPatterns.I_JUST_THANKS_TO);
        put(numberSuffix, FilterPatterns.NUMBER_SUFFIX);
        put(numberPrefix, FilterPatterns.NUMBER_PREFIX);
        put(discordLinks, FilterPatterns.DISCORD);
        put(webLinks, FilterPatterns.WEB_LINK);
        put(ips, FilterPatterns.IP_ADDR);
        put(announcers, FilterPatterns.ANNOUNCER);
        put(spammers, FilterPatterns.SPAMMER);
        put(insulters, FilterPatterns.INSULTER);
        put(greeters, FilterPatterns.GREETER);
    }};

    private static class FilterPatterns {
        private static final String[] ANNOUNCER = {
                // RusherHack b8
                "I just walked .+ feet!",
                "I just placed a .+!",
                "I just attacked .+ with a .+!",
                "I just dropped a .+!",
                "I just opened chat!",
                "I just opened my console!",
                "I just opened my GUI!",
                "I just went into full screen mode!",
                "I just paused my game!",
                "I just opened my inventory!",
                "I just looked at the player list!",
                "I just took a screen shot!",
                "I just swaped hands!",
                "I just ducked!",
                "I just changed perspectives!",
                "I just jumped!",
                "I just ate a .+!",
                "I just crafted .+ .+!",
                "I just picked up a .+!",
                "I just smelted .+ .+!",
                "I just respawned!",
                // RusherHack b11
                "I just attacked .+ with my hands",
                "I just broke a .+!",
                // WWE
                "I recently walked .+ blocks",
                "I just droped a .+ called, .+!",
                "I just placed a block called, .+!",
                "Im currently breaking a block called, .+!",
                "I just broke a block called, .+!",
                "I just opened chat!",
                "I just opened chat and typed a slash!",
                "I just paused my game!",
                "I just opened my inventory!",
                "I just looked at the player list!",
                "I just changed perspectives, now im in .+!",
                "I just crouched!",
                "I just jumped!",
                "I just attacked a entity called, .+ with a .+",
                "Im currently eatting a peice of food called, .+!",
                "Im currently using a item called, .+!",
                "I just toggled full screen mode!",
                "I just took a screen shot!",
                "I just swaped hands and now theres a .+ in my main hand and a .+ in my off hand!",
                "I just used pick block on a block called, .+!",
                "Ra just completed his blazing ark",
                "Its a new day yes it is",
                // DotGod.CC
                "I just placed .+ thanks to (http:\\/\\/)?DotGod\\.CC!",
                "I just flew .+ meters like a butterfly thanks to (http:\\/\\/)?DotGod\\.CC!",
        };

        private static final String[] SPAMMER = {
                //WWE
                "WWE Client's spammer",
                "Lol get gud",
                "Future client is bad",
                "WWE > Future",
                "WWE > Impact",
                "Default Message",
                "IKnowImEZ is a god",
                "THEREALWWEFAN231 is a god",
                "WWE Client made by IKnowImEZ/THEREALWWEFAN231",
                "WWE Client was the first public client to have Path Finder/New Chunks",
                "WWE Client was the first public client to have color signs",
                "WWE Client was the first client to have Teleport Finder",
                "WWE Client was the first client to have Tunneller & Tunneller Back Fill",
        };

        private static final String[] INSULTER = {
                // WWE
                ".+ Download WWE utility mod, Its free!",
                ".+ 4b4t is da best mintscreft serber",
                ".+ dont abouse",
                ".+ you cuck",
                ".+ https://www.youtube.com/channel/UCJGCNPEjvsCn0FKw3zso0TA",
                ".+ is my step dad",
                ".+ again daddy!",
                "dont worry .+ it happens to every one",
                ".+ dont buy future it's crap, compared to WWE!",
                "What are you, fucking gay, .+?",
                "Did you know? .+ hates you, .+",
                "You are literally 10, .+",
                ".+ finally lost their virginity, sadly they lost it to .+... yeah, that's unfortunate.",
                ".+, don't be upset, it's not like anyone cares about you, fag.",
                ".+, see that rubbish bin over there? Get your ass in it, or I'll get .+ to whoop your ass.",
                ".+, may I borrow that dirt block? that guy named .+ needs it...",
                "Yo, .+, btfo you virgin",
                "Hey .+ want to play some High School RP with me and .+?",
                ".+ is an Archon player. Why is he on here? Fucking factions player.",
                "Did you know? .+ just joined The Vortex Coalition!",
                ".+ has successfully conducted the cactus dupe and duped a itemhand!",
                ".+, are you even human? You act like my dog, holy shit.",
                ".+, you were never loved by your family.",
                "Come on .+, you hurt .+'s feelings. You meany.",
                "Stop trying to meme .+, you can't do that. kek",
                ".+, .+ is gay. Don't go near him.",
                "Whoa .+ didn't mean to offend you, .+.",
                ".+ im not pvping .+, im WWE'ing .+.",
                "Did you know? .+ just joined The Vortex Coalition!",
                ".+, are you even human? You act like my dog, holy shit.",
        };

        private static final String[] GREETER = {
                // WWE
                "Bye, Bye .+",
                "Farwell, .+",
                // Others(?)
                "See you next time, .+",
                "Catch ya later, .+",
                "Bye, .+",
                "Welcome, .+",
                "Hey, .+",
                // Vanilla MC / Essentials MC
                ".+ joined the game",
                ".+ has joined",
                ".+ joined the lobby",
                "Welcome .+",
                ".+ left the game",
        };

        private static final String[] DISCORD = {
                "discord.gg",
                "discordapp.com",
                "discord.io",
                "invite.gg",
        };

        private static final String[] NUMBER_SUFFIX = {
                ".+\\d{3,}$",
        };

        private static final String[] NUMBER_PREFIX = {
                "\\d{3,}.*$",
        };

        private static final String[] GREEN_TEXT = {
                "^>.+$",
        };

        private static final String[] WEB_LINK = {
                "http:\\/\\/",
                "https:\\/\\/",
                "www.",
        };

        private static final String[] IP_ADDR = {
                "\\b\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\:\\d{1,5}\\b",
                "\\b\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}",
                "^(?:http(?:s)?:\\/\\/)?(?:[^\\.]+\\.)?.*\\..*\\..*$",
                ".*\\..*\\:\\d{1,5}$",
        };

        private static final String[] OWNS_ME_AND_ALL = {
                "owns me and all",
        };

        private static final String[] I_JUST_THANKS_TO = {
                "i just.*thanks to",
                "i just.*using",
        };

        private static final String[] SPECIAL_BEGINNING = {
                "^[.,/?!()\\[\\]{}<>|\\-+=\\\\]", // the <> don't filter as the player name is removed when matching
        };

        private static final String[] SPECIAL_ENDING = {
                "[/@#^()\\[\\]{}<>|\\-+=\\\\]",
        };
    }

    private void sendResult(String name, String message) {
        if (showBlocked.getValue().equals(ShowBlocked.CHAT)) sendChatMessage(getChatName() + name + ": " + message);
        else if (showBlocked.getValue().equals(ShowBlocked.LOG_FILE))
            KamiMod.log.info(getChatName() + name + ": " + message);
    }
}

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt
package me.zeroeightsix.kami.module.modules.chat

import me.zeroeightsix.kami.KamiMod
import me.zeroeightsix.kami.module.Module
import me.zeroeightsix.kami.setting.Settings
import me.zeroeightsix.kami.util.text.MessageDetectionHelper
import me.zeroeightsix.kami.util.text.MessageDetectionHelper.detect
import me.zeroeightsix.kami.util.text.MessageSendHelper
import me.zeroeightsix.kami.util.text.Regexes
import me.zeroeightsix.kami.util.text.SpamFilters
import net.minecraft.util.text.TextComponentString
import net.minecraftforge.client.event.ClientChatReceivedEvent
import org.kamiblue.event.listener.listener
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern

@Module.Info(
        name = "AntiSpam",
        category = Module.Category.CHAT,
        description = "Removes spam and advertising from the chat",
        showOnArray = Module.ShowOnArray.OFF
)
object AntiSpam : Module() {
    private val mode = register(Settings.e<Mode>("Mode", Mode.REPLACE))
    private val replaceMode = register(Settings.enumBuilder(ReplaceMode::class.java, "ReplaceMode").withValue(ReplaceMode.ASTERISKS).withVisibility { mode.value == Mode.REPLACE })
    private val page = register(Settings.e<Page>("Page", Page.TYPE))

    /* Page One */
    private val discordLinks = register(Settings.booleanBuilder("Discord").withValue(true).withVisibility { page.value == Page.TYPE })
    private val slurs = register(Settings.booleanBuilder("Slurs").withValue(true).withVisibility { page.value == Page.TYPE })
    private val swears = register(Settings.booleanBuilder("Swears").withValue(false).withVisibility { page.value == Page.TYPE })
    private val automated = register(Settings.booleanBuilder("Automated").withValue(true).withVisibility { page.value == Page.TYPE })
    private val ips = register(Settings.booleanBuilder("ServerIps").withValue(true).withVisibility { page.value == Page.TYPE })
    private val specialCharEnding = register(Settings.booleanBuilder("SpecialEnding").withValue(true).withVisibility { page.value == Page.TYPE })
    private val specialCharBegin = register(Settings.booleanBuilder("SpecialBegin").withValue(true).withVisibility { page.value == Page.TYPE })
    private val greenText = register(Settings.booleanBuilder("GreenText").withValue(false).withVisibility { page.value == Page.TYPE })
    private val fancyChat = register(Settings.booleanBuilder("FancyChat").withValue(false).withVisibility { page.value == Page.TYPE })

    /* Page Two */
    private val aggressiveFiltering = register(Settings.booleanBuilder("AggressiveFiltering").withValue(true).withVisibility { page.value == Page.SETTINGS })
    private val duplicates = register(Settings.booleanBuilder("Duplicates").withValue(true).withVisibility { page.value == Page.SETTINGS })
    private val duplicatesTimeout = register(Settings.integerBuilder("DuplicatesTimeout").withValue(30).withRange(1, 600).withStep(5).withVisibility { duplicates.value && page.value == Page.SETTINGS })
    private val filterOwn = register(Settings.booleanBuilder("FilterOwn").withValue(false).withVisibility { page.value == Page.SETTINGS })
    private val filterDMs = register(Settings.booleanBuilder("FilterDMs").withValue(false).withVisibility { page.value == Page.SETTINGS })
    private val filterServer = register(Settings.booleanBuilder("FilterServer").withValue(false).withVisibility { page.value == Page.SETTINGS })
    private val showBlocked = register(Settings.enumBuilder(ShowBlocked::class.java, "ShowBlocked").withValue(ShowBlocked.LOG_FILE).withVisibility { page.value == Page.SETTINGS })

    private enum class Mode {
        REPLACE, HIDE
    }

    @Suppress("unused")
    private enum class ReplaceMode(val redaction: String) {
        REDACTED("[redacted]"), ASTERISKS("****")
    }

    private enum class Page {
        TYPE, SETTINGS
    }

    @Suppress("unused")
    private enum class ShowBlocked {
        NONE, LOG_FILE, CHAT, BOTH
    }

    private val messageHistory = ConcurrentHashMap<String, Long>()
    private val settingMap = hashMapOf(
            greenText to SpamFilters.greenText,
            specialCharBegin to SpamFilters.specialBeginning,
            specialCharEnding to SpamFilters.specialEnding,
            automated to SpamFilters.ownsMeAndAll,
            automated to SpamFilters.thanksTo,
            discordLinks to SpamFilters.discordInvite,
            ips to SpamFilters.ipAddress,
            automated to SpamFilters.announcer,
            automated to SpamFilters.spammer,
            automated to SpamFilters.insulter,
            automated to SpamFilters.greeter,
            slurs to SpamFilters.slurs,
            swears to SpamFilters.swears
    )

    init {
        listener<ClientChatReceivedEvent> { event ->
            if (mc.player == null) return@listener

            messageHistory.values.removeIf { System.currentTimeMillis() - it > 600000 }

            if (duplicates.value && checkDupes(event.message.unformattedText)) {
                event.isCanceled = true
            }

            val pattern = isSpam(event.message.unformattedText)
            if (pattern != null) { // null means no pattern found
                if (mode.value == Mode.HIDE) {
                    event.isCanceled = true
                } else if (mode.value == Mode.REPLACE) {
                    event.message = TextComponentString(sanitize(event.message.formattedText, pattern, (replaceMode.value as ReplaceMode).redaction))
                }
            }

            if (fancyChat.value) {
                val message = sanitizeFancyChat(event.message.unformattedText)
                if (message.trim { it <= ' ' }.isEmpty()) { // this should be removed if we are going for an intelligent de-fancy
                    event.message = TextComponentString(getUsername(event.message.unformattedText) + " [Fancychat]")
                }
            }
        }
    }

    override fun onDisable() {
        messageHistory.clear()
    }

    private fun sanitize(toClean: String, matcher: String, replacement: String): String {
        return if (!aggressiveFiltering.value) {
            toClean.replace("\\b$matcher|$matcher\\b".toRegex(), replacement) // only check for start or end of a word
        } else { // We might encounter the scunthorpe problem, so aggressive mode is off by default.
            toClean.replace(matcher.toRegex(), replacement)
        }
    }

    private fun isSpam(message: String): String? {
        return if (!filterOwn.value && isOwn(message)
                || MessageDetectionHelper.isDirect(!filterDMs.value, message)
                || message.detect(!filterServer.value, Regexes.QUEUE)
                || message.detect(!filterServer.value, Regexes.RESTART)) {
            null
        } else {
            detectSpam(removeUsername(message))
        }
    }

    private fun removeUsername(username: String): String {
        return username.replace("<[^>]*> ".toRegex(), "")
    }

    private fun getUsername(rawMessage: String): String? {
        val matcher = Pattern.compile("<[^>]*>", Pattern.CASE_INSENSITIVE).matcher(rawMessage)
        return if (matcher.find()) {
            matcher.group()
        } else {
            rawMessage.substring(0, rawMessage.indexOf(">")) // a bit hacky
        }
    }

    private fun detectSpam(message: String): String? {
        for ((key, value) in settingMap) {
            val pattern = findPatterns(value, message)
            if (key.value && pattern != null) {
                sendResult(key.name, message)
                return pattern
            }
        }
        return null
    }

    private fun checkDupes(message: String): Boolean {
        var isDuplicate = false

        if (messageHistory.containsKey(message) && (System.currentTimeMillis() - messageHistory[message]!!) / 1000 < duplicatesTimeout.value) isDuplicate = true
        messageHistory[message] = System.currentTimeMillis()

        if (isDuplicate) {
            sendResult("Duplicate", message)
        }
        return isDuplicate
    }


    private fun isOwn(message: String): Boolean {
        /* mc.player is null when the module is being registered, so this matcher isn't added alongside the other FilterPatterns */
        val ownFilter = "^<" + mc.player.name + "> "
        return Pattern.compile(ownFilter, Pattern.CASE_INSENSITIVE).matcher(message).find()
    }

    private fun findPatterns(patterns: Array<String>, string: String): String? {
        var cString = string
        cString = cString.replace("<[^>]*> ".toRegex(), "") // remove username first
        for (pattern in patterns) {
            if (Pattern.compile(pattern, Pattern.CASE_INSENSITIVE).matcher(cString).find()) {
                return pattern
            }
        }
        return null
    }

    private fun sanitizeFancyChat(toClean: String): String {
        // this has the potential to be intelligent and convert to ascii instead of just delete
        return toClean.replace("[^\\u0000-\\u007F]".toRegex(), "")
    }

    private fun sendResult(name: String, message: String) {
        if (showBlocked.value == ShowBlocked.CHAT || showBlocked.value == ShowBlocked.BOTH) MessageSendHelper.sendChatMessage("$chatName $name: $message")
        if (showBlocked.value == ShowBlocked.LOG_FILE || showBlocked.value == ShowBlocked.BOTH) KamiMod.LOG.info("$chatName $name: $message")
    }
}
